plugins/chat/spec/services/chat/auto_join_channels_spec.rb

```diff
@@ -0,0 +1,359 @@
+    # Had to noop all the event handlers because of the "DiscourseEvent registrations were not cleaned up" spec
+    before do
+      %i[
+        site_setting_changed
+        group_destroyed
+        user_seen
+        user_confirmed_email
+        user_added_to_group
+        user_removed_from_group
+        category_updated
+      ].each { |event| DiscourseEvent.events[event] = [noop] * DiscourseEvent.events[event].size }
+    end
+
+    context "when chat is disabled" do
+      before { SiteSetting.chat_enabled = false }
+
+      it { is_expected.to fail_a_policy(:chat_enabled?) }
+    end
+
+    context "when chat is enabled" do
+      before { SiteSetting.chat_enabled = true }
+
+      let(:trust_level) { 1 } # SiteSetting.chat_allowed_groups defaults to admins, moderators, and TL1 users
+      let(:last_seen_at) { 5.minutes.ago } # Users must have been seen "recently" to be auto-joined to a channel
+
+      fab!(:public_category) { Fabricate(:category) }
+      fab!(:private_category) { Fabricate(:category, read_restricted: true) }
+
+      fab!(:private_group_readonly) { Fabricate(:group) }
+      fab!(:private_group_create_post) { Fabricate(:group) }
+      fab!(:private_group_full) { Fabricate(:group) }
+
+      fab!(:private_category_group_readonly) do
+        Fabricate(
+          :category_group,
+          category: private_category,
+          group: private_group_readonly,
+          permission_type: CategoryGroup.permission_types[:readonly],
+        )
+      end
+
+      fab!(:private_category_group_create_post) do
+        Fabricate(
+          :category_group,
+          category: private_category,
+          group: private_group_create_post,
+          permission_type: CategoryGroup.permission_types[:create_post],
+        )
+      end
+
+      fab!(:private_category_group_full) do
+        Fabricate(
+          :category_group,
+          category: private_category,
+          group: private_group_full,
+          permission_type: CategoryGroup.permission_types[:full],
+        )
+      end
+
+      context "with a non-auto joinable public channel" do
+        fab!(:non_auto_joinable_public_channel) do
+          Fabricate(:chat_channel, chatable: public_category)
+        end
+
+        it "doesn't automatically join users" do
+          Fabricate(:user, trust_level:, last_seen_at:)
+
+          expect { job }.not_to change { Chat::UserChatChannelMembership.count }.from(0)
+        end
+      end
+
+      context "with an auto joinable public channel" do
+        fab!(:auto_joinable_public_channel) do
+          Fabricate(:chat_channel, chatable: public_category, auto_join_users: true)
+        end
+
+        it "automatically joins users" do
+          2.times { Fabricate(:user, trust_level:, last_seen_at:) }
+
+          expect { job }.to change { Chat::UserChatChannelMembership.count }.from(0).to(2)
+        end
+
+        it "doesn't automatically join users who have chat disabled" do
```

Comment:
We could probably wrap examples like this one into a context (saying something like `when some users have chat disabled`), wdyt?

Reply:
Does it _really_ need a context? I mean, all the other test cases have the same "pattern" - setup line(s) and then expect(s)

Reply:
That’s what context is for :sweat_smile: At least could we group all the happy path examples together?

Reply:
Yes, sorry, I was picking up on the other test cases part :sweat_smile: But it seems you answered my other comment.